airflow/executors/knative_executor.py

```diff
@@ -0,0 +1,210 @@
+            return resp
+
+
+class KnativeRequestLoop(multiprocessing.Process, LoggingMixin):
+    """
+
+    This class asynchronously pulls tasks from the KnativeExecutor and runs them as coroutines using asyncio
+
+    """
+    def __init__(self,
+                 task_pipe,
+                 result_pipe,
+                 host_header,
+                 host,
+                 ):
+
+        super().__init__()
+        self.host = host
+        self.task_pipe = task_pipe
+        self.result_pipe = result_pipe
+        self.host_header = host_header
+
+    def recieve_and_execute(self, loop):
+        """
+        This function receives tasks from task_pipe and executes them in FIFO order.
+        @param loop:
+        """
+        key = self.task_pipe.recv()
+        loop.create_task(self.execute_work(key=key))
+
+    def run(self):
+        """
+        Starts loop and prepares pipe reader.
+        """
+        loop = asyncio.get_event_loop()
+        loop.add_reader(fd=self.task_pipe, callback=functools.partial(self.recieve_and_execute, loop))
+        loop.run_forever()
+
+    async def execute_work(self, key):
+        """
+        Executes command received and stores result state in queue.
+        :param task_instance:
+        :param key: the key to identify the TI
+        :type key: tuple(dag_id, task_id, execution_date)
+        :param command: the command to execute
+        :type command: str
+        """
+        if key is None:
+            return
+        self.log.info("%s running %s", self.__class__.__name__, key)
+        (dag_id, task_id, execution_date, _) = key
+        try:
+            future = make_request_async(
+                task_id,
+                dag_id,
+                execution_date,
+                host=self.host, log=self.log, host_header=self.host_header)
+            resp = await future
+            if resp.status != 200:
+                state = State.FAILED
+                self.log.error("Failed to execute task %s.", str(resp.text))
+                self.result_pipe.send((key, state))
+            else:
+                self.log.info("assuming final state has been handled")
+                self.result_pipe.send((key, None))
+
+        except asyncio.InvalidStateError as e:
+            state = State.FAILED
+            self.log.error("Failed to execute task %s.", str(e))
+            self.result_pipe.send((key, state))
+
+
+class KnativeExecutor(BaseExecutor):
+    """
+    KnativeExecutor executes tasks locally in parallel. It uses the
+    multiprocessing Python library and queues to parallelize the execution
+    of tasks.
+    """
+    task_queue = None
+    result_queue = None
+    result_pipe = None
+
+    def terminate(self):
+        pass
+
+    @cached_property
+    def kube_executor(self) -> KubernetesExecutor:
+        """
+        lazily loads KubernetesExecutor
+        @return:
+        """
+        self.kube_executor_initialized = True
+        return KubernetesExecutor()
```

Comment:
Can you describe one more time why we use this like that?

Reply:
We don't want to start a KubernetesExecutor unnecessarily. This function essentially acts as a lazy loader s.t. if a user requests an isolated pod (either due to a very long-running task or because they want it fully isolated) it will start a KubernetesExecutor.